compiler/rustc_data_structures/src/functor.rs

```diff
@@ -36,38 +36,43 @@ impl<T> IdFunctor for Vec<T> {
     type Inner = T;
 
     #[inline]
-    fn try_map_id<F, E>(mut self, mut f: F) -> Result<Self, E>
+    fn try_map_id<F, E>(self, mut f: F) -> Result<Self, E>
     where
         F: FnMut(Self::Inner) -> Result<Self::Inner, E>,
     {
-        // FIXME: We don't really care about panics here and leak
-        // far more than we should, but that should be fine for now.
-        let len = self.len();
-        unsafe {
-            self.set_len(0);
-            let start = self.as_mut_ptr();
-            for i in 0..len {
-                let p = start.add(i);
-                match f(p.read()) {
-                    Ok(val) => p.write(val),
-                    Err(err) => {
-                        // drop all other elements in self
-                        // (current element was "moved" into the call to f)
-                        for j in (0..i).chain(i + 1..len) {
-                            start.add(j).drop_in_place();
-                        }
+        struct HoleVec<T> {
+            vec: Vec<mem::ManuallyDrop<T>>,
+            hole: Option<usize>,
+        }
 
-                        // returning will drop self, releasing the allocation
-                        // (len is 0 so elements will not be re-dropped)
-                        return Err(err);
+        impl<T> Drop for HoleVec<T> {
+            fn drop(&mut self) {
+                unsafe {
+                    let len = self.vec.len();
+                    let hole = self.hole.unwrap_or(len);
+                    for index in (0..hole).chain(1 + hole..len) {
+                        mem::ManuallyDrop::drop(self.vec.get_unchecked_mut(index));
                     }
```

Comment:
Consider splitting this into 2 distinct for loops. `chain` is sadly not as 0-cost as one would like. This is also quite non-trivial and I would be much more comfortable with e.g.

```rust
for (idx, val) = self.vec.iter_mut().enumerate() {
    if idx != hole { ManuallyDrop::drop(val) } 
}
```

Reply:
I wanted to avoid the comparison and branch within the loop, but I guess this drop-path is pretty cold anyway and it's likely very immaterial.